File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;
int answer=0;
void DFS(vector<int> &numbers, int target, int cnt, int sum)
{
    if(cnt==numbers.size())
    {
        if(target==sum) answer++;
        return;
    }
    DFS(numbers,target,cnt+1,sum+numbers[cnt]);
    DFS(numbers,target,cnt+1,sum-numbers[cnt]);
}
int solution(vector<int> numbers, int target) {
    DFS(numbers,target,0,0);
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    private int[] numbers;
    private int target;
    private int answer;
    
    public int solution(int[] numbers, int target)
    {
        this.numbers = numbers;
        this.target = target;
        
        DFS(0,0);
        return answer;
    }
    
    public void DFS(int sum, int depth) {
        if(depth == numbers.length) {
            if(sum == target) {
                answer++;
            }
            return;
        }
        
        DFS(sum + numbers[depth], depth+1);
        DFS(sum - numbers[depth], depth+1);
    }
}